### PR TITLE
Sentry stats lib + cli commands

### PR DESCRIPTION
**Pull request by danielSanchezQ**
This change works on top of #21 
It adds the stats command to the cli and the utilities for it.

### CODE REVIEW

File: src/bin/cli/logs/sentry/download.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    }
+}
+
+fn filter_logs_by_date(
+    logs: impl Iterator<Item = RawLog>,
+    dates: DateFilter,
+) -> impl Iterator<Item = RawLog> {
+    let DateFilter { from, to } = dates;
+
+    logs.take_while(move |l| {
```
Comment (Zeegomo) [inline src/bin/cli/logs/sentry/download.rs:93]: Does this work though? I think `take_while` stops after the first `false` value, so if the first log is before `from_date` no logs will be returned even though there are valid ones in the specified window

File: src/logs/sentry.rs [hunk 2]
```diff
@@ -94,119 +104,183 @@ impl IntoIterator for LazySentryLogs {
...
+}
+
+pub struct RegexMatch {
+    key: String,
+    re: Regex,
+    total_checked: usize,
+    pub matches: usize,
+}
+
+pub enum SentryLogsStatChecker {
```
Comment (Zeegomo) [inline src/logs/sentry.rs:132]: I think trait objects could be used without the need to have an enum, but that's just personal preference

File: src/logs/sentry.rs [hunk 3]
```diff
@@ -94,119 +104,183 @@ impl IntoIterator for LazySentryLogs {
...
-            choice: log.choice,
-            fragment_id: log.fragment_id,
+impl Stat for SuccessfulScan {
+    fn check_raw_log(&mut self, log: &RawLog) {
+        if raw_log_message_starts_with(log, REGISTERED_MESSAGE) {
+            self.total += 1;
         }
     }
+
+    fn report(&self) {
```
Comment (Zeegomo) [inline src/logs/sentry.rs:164]: Maybe pass a buffer or similar like `std::fmt::Formatter` to be a little more flexible if we don't want to pollute stdout?

File: src/bin/cli/logs/sentry/download.rs (hunk 1 above)
Comment (danielSanchezQ) [reply src/bin/cli/logs/sentry/download.rs:93]: Uhm, what do you mean? logs come sorted by date. `from_date` is the earliest one of them.

File: src/logs/sentry.rs (hunk 3 above)
Comment (danielSanchezQ) [reply src/logs/sentry.rs:164]: Aah, yeah, that is indeed better though. Will do.

File: src/logs/sentry.rs (hunk 2 above)
Comment (danielSanchezQ) [reply src/logs/sentry.rs:132]: Yeah, but the idea behind this was to be able to access some public internal attributes if needed. Which you can't do with traits directly iirc.

File: src/bin/cli/logs/sentry/download.rs (hunk 1 above)
Comment (Zeegomo) [reply src/bin/cli/logs/sentry/download.rs:93]: Ah sorry, I thought they were sorted with increasing dates, but they are reversed

File: src/logs/sentry.rs [hunk 4]
```diff
@@ -161,8 +163,8 @@ impl Stat for SuccessfulScan {
         }
     }
 
-    fn report(&self) {
-        println!("Total successful scans: {}", self.total);
+    fn report(&self, formatter: &mut std::fmt::Formatter) -> fmt::Result {
+        formatter.write_fmt(format_args!("Total successful scans: {}", self.total))
```
Comment (Zeegomo) [inline src/logs/sentry.rs:167]: Why not the `write/writeln` macro?

Comment (danielSanchezQ) [reply src/logs/sentry.rs:167]: Just used to do it like that, but totally true.

**Review by dkijania — APPROVED** [review]

**Comment by dkijania** [comment]
Looking good. thanks!

**Pull request merged**